### PR TITLE
tc actions support

### PR DESCRIPTION
**Pull request by daniel-noland**
# Support for [tc-actions](https://man7.org/linux/man-pages/man8/tc-actions.8.html)

Add the building blocks for tc-actions support.

This PR is in the same spirit as my (in progress) work on florianl/go-tc#145

I broke out this functionality from my other PR #111 based on conversation and feedback from @cathay4t.

## Summary of feature

This is easiest to explain by way of iproute2.

`tc` allows actions to be created independently of filters.
Once created, these actions may then

1. be associated with _zero or more_ filters,
2. live updated (and updates will be seen by all filters using that action),
3. be deleted (only after all filters using that action have been deleted).

For example, consider the following :

```bash
for i in x y z; do
  ip link add dev "$i" type dummy
  tc qdisc add dev "$i" clsact
done

tc actions add action mirred egress redirect dev y
tc actions add action gact drop 
```

At this point, we could

1. list the `mirred` actions

   ```bash
   $ tc actions list action mirred      
   total acts 1

           action order 0: mirred (Egress Redirect to device y) stolen
           index 1 ref 1 bind 0
           not_in_hw
           used_hw_stats disabled
   ```

2. list the `gact` actions

    ```bash
    $ tc actions list action gact
    total acts 1
    
            action order 0: gact action drop
             random type none pass val 0
             index 1 ref 1 bind 0
            not_in_hw
            used_hw_stats disabled
    ```

3. create any number of filters using either or both of these actions by index

    ```bash
    tc filter add dev x ingress pref 1000 proto ip flower dst_ip 8.8.8.8 action mirred index 1
    tc filter add dev z ingress pref 1000 proto ip flower dst_ip 8.8.8.8 action mirred index 1 action gact index 1   
    ```

4. display those filters as normal (with per-action statistics)

   ```bash
   $ tc -s filter show dev z ingress
   filter protocol ip pref 1000 flower chain 0
   filter protocol ip pref 1000 flower chain 0 handle 0x1
   eth_type ipv4
   dst_ip 8.8.8.8
   not_in_hw
          action order 1: mirred (Egress Redirect to device y) stolen
          index 1 ref 3 bind 2 installed 599 sec used 599 sec
          Action statistics:
          Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
          backlog 0b 0p requeues 0
   
          action order 2: gact action drop
           random type none pass val 0
           index 1 ref 2 bind 1 installed 599 sec used 599 sec
          Action statistics:
          Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
          backlog 0b 0p requeues 0
   ```

5. centrally update those actions (e.g., change `drop` to `pass`)

   ```bash
   $ tc actions change action gact pass index 1
   $ tc -s filter show dev z ingress 
   filter protocol ip pref 1000 flower chain 0
   filter protocol ip pref 1000 flower chain 0 handle 0x1
   eth_type ipv4
   dst_ip 8.8.8.8
   not_in_hw
   action order 1: mirred (Egress Redirect to device y) stolen
   index 1 ref 3 bind 2 installed 838 sec used 838 sec
   Action statistics:
   Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
   backlog 0b 0p requeues 0
   
           action order 2: gact action pass
            random type none pass val 0
            index 1 ref 2 bind 1 installed 838 sec used 838 sec
           Action statistics:
           Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
           backlog 0b 0p requeues 0
   ```
   
6. attempts to delete those actions while in use will be rejected (albeit with a buggy error message from `iproute2`/`tc`)

   ```bash
   $ tc actions delete action gact index 1
   Error: Failed to delete TC action.
   We have an error talking to the kernel
   Command "action" is unknown, try "tc actions help"
   ```
   
7. Removing all filters that use an action will allow the action to be deleted

   ```bash
   $ tc filter del dev z ingress pref 1000
   $ tc actions delete action gact index 1
   $ tc filter del dev x ingress pref 1000
   $ tc actions delete action mirred index 1
   ```

### CODE REVIEW

**Comment by daniel-noland** [comment]
I'm moving this out of draft state now because I think it can now be meaningfully reviewed.

I will work on squashing the remaining tiny TODO involving documentation, making sure all the doc links work correctly, and the other notes I left for myself.  I will be very surprised if addressing any of those significantly change the code.

As things stand I

1. Wrote more documentation for all of the `tc-actions` functionality.
2. Added support for `tc-mirred`

in addition to what is described in the issue.

If you like I will break this into three commits to account for that (or even three PRs if you feel it is necessary).

**Comment by daniel-noland** [comment]
I also made a slight amendment to the testing pattern by 

1. pulling in the hex library as a dev dependency.
2. adding additional tests as per my thoughts in #121

I am happy to adapt that back into the existing testing pattern with hard coded `vec![...]` statements and or remove the "parse back" style tests if you don't think they belong here.

File: src/tc/actions/action.rs [hunk 1]
```diff
@@ -49,62 +56,53 @@ impl Nla for TcAction {
...
-                        .context("failed to parse action nla")?,
-                ),
-            });
-        }
+                if kinds.len() > 1 {
+                    return Err(DecodeError::from("Duplicate TCA_ACT_KIND"));
+                }
+                kinds[0].clone()
+            }
+            Err(e) => return Err(DecodeError::from(e.to_string())),
```
Comment (wllenyj) [inline src/tc/actions/action.rs:97]: Does it affect unsupported cases?

File: src/message.rs [hunk 2]
```diff
@@ -291,6 +291,21 @@ impl<'a, T: AsRef<[u8]> + ?Sized>
                 }
             }
 
+            RTM_NEWACTION | RTM_DELACTION | RTM_GETACTION => {
+                let err = "invalid tc action message";
+                let msg = TcActionMessage::parse(
+                    &TcActionMessageBuffer::new_checked(&buf.inner())
+                        .context(err)?,
+                )
+                .context(err)?;
```
Comment (wllenyj) [inline src/message.rs:300]: Is `context` required?

**Review by wllenyj — APPROVED** [review]
Thanks for your documentation.
The unit tests is nice too.

File: src/message.rs (hunk 2 above)
Comment (daniel-noland) [reply src/message.rs:300]: good catch.  No idea how that got there.  Will fix

File: src/tc/actions/action.rs (hunk 1 above)
Comment (daniel-noland) [reply src/tc/actions/action.rs:97]: I am not 100% sure what you mean.

That said, I think this bit of code is kinda messy.  I will think about a better and more clear way to express this logic.

**Comment by daniel-noland** [comment]
> Thanks for your documentation. The unit tests is nice too.

Thank you for the review :smile:

I will write fixups based on your comments.

If you think that flavor of unit tests is helpful (as I do, see #121) then we may also consider something like [proptest](https://github.com/proptest-rs/proptest) or [bolero](https://github.com/camshaft/bolero) to really expand on that "parse-back" strategy.
We could even use bolero+[kani](https://github.com/model-checking/kani) if we wanna get really hardcore, but things like that are best done one step at a time. 

It may also be worth looking at the rather unfortunate state of the higher level API this PR results in (see rust-netlink/rtnetlink#64) before we agree to merge this.
If changes to this PR are needed to smooth out the rtnetlink api then that is best done before anybody (including rtnetlink) starts depending on this change set.
Especially if we are going to try to 1.0 this library this year :smile: (something I really look forward to)

[This segment of test code is especially unfortunate](https://github.com/rust-netlink/rtnetlink/blob/ca1a7983e66d815fef776b7b8f32cf84444598a8/src/traffic_control/test.rs#L474-L548)

[This segment of test code is almost as sad](https://github.com/rust-netlink/rtnetlink/blob/ca1a7983e66d815fef776b7b8f32cf84444598a8/src/traffic_control/test.rs#L656-L717)

I have thoughts about what can be done to fix this.
I will write them up shortly.

Thanks again :tada:

File: src/message.rs (hunk 2 above)
Comment (daniel-noland) [reply src/message.rs:300]: It looks like I was just following an existing pattern when I did this?

If we change this pattern here we should likely change this pattern for all the other arms to remain consistent.

Thoughts?

File: src/tc/actions/action.rs (hunk 1 above)
Comment (wllenyj) [reply src/tc/actions/action.rs:97]: Sorry, my mistake. 
Due to the diff is not match to the source code. The souce of parse move to `TcActionAttribute::parse_with_param`.

File: src/message.rs (hunk 2 above)
Comment (wllenyj) [reply src/message.rs:300]: It's ok to keep origin pattern.
But the `context` is used to trace error path. I haven't tried what `context(err)` behaves like.

Comment (daniel-noland) [reply src/message.rs:300]: Ok.  Then I suggest we leave it as is for now.

File: src/tc/actions/action.rs (hunk 1 above)
Comment (daniel-noland) [reply src/tc/actions/action.rs:97]: Ok no worries.

Mark resolved?

File: src/tc/actions/action.rs [hunk 3]
```diff
@@ -165,7 +203,17 @@ impl Nla for TcActionAttribute {
...
         match self {
             Self::Kind(_) => TCA_ACT_KIND,
-            Self::Options(_) => TCA_ACT_OPTIONS,
+            Self::Options(opts) => {
+                // NOTE: the kernel simply doesn't consistently use the nested
+                // flag based on captured messages.
+                // This is heuristically trying to match the kernel's behavior
+                // but may not be correct.
+                if opts.len() == 1 {
+                    TCA_ACT_OPTIONS | NLA_F_NESTED
```
Comment (cathay4t) [inline src/tc/actions/action.rs:212]: In kernel we have 

```c
static const struct nla_policy tcf_action_policy[TCA_ACT_MAX + 1] = {
        [TCA_ACT_KIND]          = { .type = NLA_STRING },
        [TCA_ACT_INDEX]         = { .type = NLA_U32 },
        [TCA_ACT_COOKIE]        = { .type = NLA_BINARY,
                                    .len = TC_COOKIE_MAX_SIZE },
        [TCA_ACT_OPTIONS]       = { .type = NLA_NESTED },
```

And `nest = nla_nest_start_noflag(skb, TCA_ACT_OPTIONS);`.

So it should always carry with `NLA_F_NESTED`.

File: src/tc/actions/message.rs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        ///
+        /// If set, this flag enables more than TCA_ACT_MAX_PRIO actions in a single
+        /// actions listing operation.
+        ///
+        /// > TCA_ACT_FLAG_LARGE_DUMP_ON user->kernel to request for larger than
+        ///   TCA_ACT_MAX_PRIO actions in a dump.
+        ///   All dump responses will contain the number of actions being dumped
+        ///   stored in for user app's consumption in TCA_ROOT_COUNT
+        ///
+        /// [`rtnetlink.h`]: https://github.com/iproute2/iproute2/blob/89210b9ec1c445ae963d181b5816d12a0cdafbb6/include/uapi/linux/rtnetlink.h#L803-L806
```
Comment (cathay4t) [inline src/tc/actions/message.rs:41]: Those link might not accessible in the future. No need to mention it.

iproute2 might be the de-factor netlink client, but I prefer explains stuff using kernel codes or function name.

File: src/tc/actions/action.rs (hunk 1 above)
Comment (wllenyj) [reply src/tc/actions/action.rs:97]: LGTM

File: src/tc/tests/action_nat.rs [hunk 5]
```diff
@@ -1,164 +0,0 @@
-// SPDX-License-Identifier: MIT
```
Comment (cathay4t) [inline src/tc/tests/action_nat.rs:1]: Why this file is removed instead of amended?

**Review by cathay4t — CHANGES_REQUESTED** [review]
Besides the lints I mentioned, all looks good to me.

We have document in README.md for capture netlink data and convert it into Vec<u8>, which could be better than decode a long string.

**Comment by cathay4t** [comment]
If you find it hard to capture netlink package as Vec<u8>, please let me know, I am happy to amend(and also validate) your test codes.

**Comment by daniel-noland** [comment]
> If you find it hard to capture netlink package as Vec, please let me know, I am happy to amend(and also validate) your test codes.

I can capture as a `Vec` if that is better.  It does remove the dev-dependency on hex.  Will change

File: src/tc/tests/action_nat.rs (hunk 5 above)
Comment (daniel-noland) [reply src/tc/tests/action_nat.rs:1]: Hmm, I thought I moved the contents of this file into the same file as the other nat tests.

Likely that change was messed up in a rebase.  Will fix.

File: src/tc/actions/message.rs (hunk 4 above)
Comment (daniel-noland) [reply src/tc/actions/message.rs:41]: Ok.  Will fix.

File: src/tc/actions/action.rs (hunk 3 above)
Comment (daniel-noland) [reply src/tc/actions/action.rs:212]: Ah, it must be iproute2 just being wrong sometimes.  I will change our code to always set NLA_F_NESTED

**Comment by daniel-noland** [comment]
Feedback responses posted.

Summary of changes:

1. adjust use of `NLA_F_NESTED`
flag in both tests and emitted messages
to reflect the kernel's stated usage
(i.e. that `NLA_F_NESTED` be set for the tc-actions options).

   This required adjusting the tests which seem to have captured iproute2's incorrectly formed messages (iproute2 does not always set the `NLA_F_NESTED` flag)

2. Remove the use of hex as a dev dependency
and use a const slice of `u8` instead as per @cathay4t request

3. remove links to iproute2's github as per @cathay4t request

4. trivial documentation cleanup

Will mark as ready for next review pass.
Thanks for review.

File: src/tc/tests/action_nat.rs (hunk 5 above)
Comment (daniel-noland) [reply src/tc/tests/action_nat.rs:1]: The test in this file has been moved now.  It lives with the rest of the NAT tests

**Comment by daniel-noland** [comment]
Any update on this?

File: src/message.rs (hunk 2 above)
Comment (cathay4t) [reply src/message.rs:300]: The use of anyhow need to be debated in another PR  or issue. I don't like anyhow, but let's follow the precedent

**Review by cathay4t — APPROVED** [review]
Nicely done!

**Pull request merged**